ann_benchmarks/datasets.py

```diff
@@ -311,41 +353,29 @@ def kosarak(out_fn):
     url = 'http://fimi.uantwerpen.be/data/%s' % local_fn
     download(url, local_fn)
 
+    X = []
+    dimension = 0
     with gzip.open('kosarak.dat.gz', 'r') as f:
         content = f.readlines()
         # preprocess data to find sets with more than 20 elements
         # keep track of used ids for reenumeration
-        ids = {}
-        next_id = 0
-        cnt = 0
         for line in content:
             if len(line.split()) >= min_elements:
-                cnt += 1
-                for x in line.split():
-                    if int(x) not in ids:
-                        ids[int(x)] = next_id
-                        next_id += 1
-
-    X = numpy.zeros((cnt, len(ids)), dtype=numpy.bool)
-    i = 0
-    for line in content:
-        if len(line.split()) >= min_elements:
-            for x in line.split():
-                X[i][ids[int(x)]] = 1
-            i += 1
-
-    X_train, X_test = train_test_split(numpy.array(X), test_size=500)
-    write_output(X_train, X_test, out_fn, 'jaccard', 'bit')
+                X.append(list(map(int, line.split())))
+                dimension = max(dimension, max(X[-1]) + 1)
+
+    X_train, X_test = train_test_split(numpy.array(X), test_size=500, dimension=dimension)
+    write_sparse_output(X_train, X_test, out_fn, 'jaccard', dimension)
```

Comment:
Note that the original code only included sets that include at least 20 elements to make sure that queries are somewhat interesting.

Reply:
The diff is not very explicit, but it's still here, see https://github.com/erikbern/ann-benchmarks/pull/235/files/72019680f36a2b80856bdc3fd7e6584c49930b0b#diff-6a33a2ec7e432a5d84bba3fee0f59ec57fc44f3786a8fe1c75b6a3f617270ee2R363

Reply:
Ah, sorry. I missed that part in the diff :-)